### PR TITLE
[WIP] capture: do not close tmpfile buffer, but redirect

### PR DESCRIPTION
**Pull request by blueyed**
Fixes https://github.com/pytest-dev/pytest/issues/5502

TODO:

- [ ] changelog
- [x] currently might fail with pytest's own suite: OSError: [Errno 24] Too many open files - addressed via tmpfile-reuse
 - ~~[ ] after https://github.com/pytest-dev/pytest/pull/6036~~

### CODE REVIEW

**Comment by blueyed** [comment]
`lsof_check` would need to handle this: https://github.com/pytest-dev/pytest/blob/df4e3a68c7dc8b3d6b3ba6b78404b24f161c68d9/testing/test_capture.py#L902-L914

But just running all `atexit` handlers there is bad..

**Comment by blueyed** [comment]
Strongly feel like not adding boilerplate (via https://github.com/pytest-dev/pytest/pull/6036), so feel free to pick this up - would be a great fix.

**Comment by blueyed** [comment]
Ok, it's not fair to dump work because of stupid GIFs.

**Comment by RonnyPfannschmidt** [comment]
this seems icky to me, imho we  shouldnt make pytest internals that much worse to work around other peoples broken setups

**Comment by blueyed** [comment]
> we shouldnt make pytest internals that much worse to work around other peoples broken setups

It is not really much worse, and has e.g. the benefit to re-use tmpfiles, which seems like a good idea in general to me.

**Comment by RonnyPfannschmidt** [comment]
upon closer reading its indeed not that bad, but it gets trickier and trickier to follow

i believe by taking in py.io.capture we might be able to simplify it overall + perhaps even enable stdio output as views into the  files to safe ram on larger test-suites as follow up

**Comment by blueyed** [comment]
> i believe by taking in py.io.capture

I think that's not wanted in general.

**Comment by nicoddemus** [comment]
> upon closer reading its indeed not that bad, but it gets trickier and trickier to follow

I agree at first glance seems complicated, but I have some suggestions which I believe can improve the code and leave it more readable (I suspect this is just a quick first version to see the reception and if it would work).

File: src/_pytest/capture.py [hunk 1]
```diff
@@ -77,19 +81,29 @@ def __init__(self, method):
         self._method = method
         self._global_capturing = None
         self._current_item = None
+        self._atexit_funcs = []  # type: List[Callable]
+        atexit.register(self._atexit_run)
```
Comment (nicoddemus) [inline src/_pytest/capture.py]: I think the tmpfile management can be isolated in a separate class, which would greatly simplify the code:

```python
class Streams(IntEnum):

    STDIN = 0
    STDOUT = 1
    STDERR = 2


@attr.s
class TemporaryFilePool:

    _tmpfiles = attr.ib(default=attr.factory(dict))

    def obtain_tmpfile(self, stream_id):
        try:
            tmpfile = self._tmpfiles[stream_id]            
        except KeyError:
            if stream_id == Streams.STDIN:
                tmpfile = open(os.devnull, "r")
            else:                
                f = TemporaryFile()
                with f:
                    tmpfile = safe_text_dupfile(f, mode="wb+")
            self._tmpfiles[stream_id] = tmpfile
        
        assert not tmpfile.closed
        return tmpfile

    def close(self):
        for f in self._tmpfiles.values():
            f.close()
        self._tmpfiles.clear()
```

We should then make `CaptureManager` create and maintain the `tmpfile_pool`, and pass it around to `FDCaptureBinary`. `CaptureManager` should then register the `atexit` handle itself, so these 3 lines would become:

```python
self._tmpfile_pool = TemporaryFilePool()
atexit.register(self._tmpfile_pool)
```

This design also allows us to change the pool implementation, so we could even possibly control it via a config variable and get back the old behavior (I'm not actually suggesting it, just mentioning that this is then possible).

File: src/_pytest/capture.py [hunk 2]
```diff
@@ -77,19 +81,29 @@ def __init__(self, method):
...
+        self._atexit_funcs.append(func)
+
+    def _atexit_run(self):
+        for func in self._atexit_funcs:
+            func()
+
     def _getcapture(self, method):
         if method == "fd":
-            return MultiCapture(out=True, err=True, Capture=FDCapture)
+            return MultiCapture(out=True, err=True, Capture=FDCapture, capman=self)
```
Comment (nicoddemus) [inline src/_pytest/capture.py]: Here we would pass `self._tmpfile_pool` to the `MultiCapture` objects.

File: src/_pytest/capture.py [hunk 3]
```diff
@@ -551,15 +573,33 @@ def __init__(self, targetfd, tmpfile=None):
             self.done = self._done
             if targetfd == 0:
```
Comment (nicoddemus) [inline src/_pytest/capture.py]: I think this entire `if/else` can now be just:

```python
tmpfile = self._tmpfile_pool.obtain_tmpfile(self.targetfd)
```

**Comment by nicoddemus** [comment]
I also wonder if we should force that all handles are closed with `testdir`?

**Comment by blueyed** [comment]
@nicoddemus 
Thanks for the review/suggestions.
Looking forward to implement them soon.
btw: I've experimented with having stdout and stderr duplicated to a third output, so that they can be displayed in order (which is useful e.g. when dropping into pdb).  Should be considered to be supported here likely.

File: src/_pytest/capture.py (hunk 2 above)
Comment (blueyed) [reply src/_pytest/capture.py]: I think passing `capman` is more universal.
(IIRC I am using this when auto-suspending on readin from stdin)

**Comment by blueyed** [comment]
> I also wonder if we should force that all handles are closed with `testdir`?

What do you mean?
Ensuring that every test gets new ones?

File: src/_pytest/capture.py (hunk 2 above)
Comment (nicoddemus) [reply src/_pytest/capture.py]: more universal means more coupling here... I prefer to pass only the necessary. I also suggest making it mandatory, so we don't have to check for its existense first (might require a few test changes I suppose).

**Comment by nicoddemus** [comment]
> btw: I've experimented with having stdout and stderr duplicated to a third output, so that they can be displayed in order (which is useful e.g. when dropping into pdb).]

I like the idea of displaying them order, others are not so keen. But to clarify, do you mean create a new stream (stdouterr) in *addition* to stdout/stderr capture, or replacing them?

**Comment by blueyed** [comment]
> I like the idea of displaying them order, others are not so keen.

Haven't heard other opinions.

> But to clarify, do you mean create a new stream (stdouterr) in _addition_ to stdout/stderr capture, or replacing them?

It would be a seperate one.
I'd like to use it for e.g. "suspend-on-stdin".
In general I could imagine to having a setting which would allow to have "captured output" instead of "captured stdout" and "captured stderr".  But with all three streams it might still be useful to show/handle stderr explicitly.

**Comment by nicoddemus** [comment]
> It would be a seperate one.
> I'd like to use it for e.g. "suspend-on-stdin".

Hmm not sure what you mean here, can you please clarify?

Anyway I really like the "captured output" idea, but this seems like the topic for a separate PR/issue.

> What do you mean?
> Ensuring that every test gets new ones?

Yes, but I was just wondering, not sure if this is a good idea or not. 😁

File: src/_pytest/capture.py (hunk 1 above)
Comment (nicoddemus) [reply src/_pytest/capture.py]: Oh I also realize now that we don't respect `basetemp` when creating `TemporaryFiles`...

**Comment by Sweenu** [comment]
Is this abandoned?

**Comment by wanam** [comment]
@blueyed Any reason why you closed this merge request?

**Comment by blueyed** [comment]
I'm not planning to work on this during the next time here.
Feel free to pick it up yourself already, of course.

**Pull request closed without merging**